File: src/engine/commands/talkcommand.cpp
#include "talkcommand.h"

#include "util.h"


#define super Command

TalkCommand::TalkCommand(QObject *parent) :
    super(parent) {

    setDescription("Talk to a specific character who's in the same area. Only "
                   "this character will hear you.\n"
                   "\n"
                   "Example: talk earl Hey Earl, how are you?");
}

TalkCommand::~TalkCommand() {
}

void TalkCommand::execute(Player *player, const QString &command) {

    super::prepareExecute(player, command);

    takeWord("to");

    GameObjectPtr character = takeObject(currentRoom()->characters());
    if (!requireSome(character, "Talk to who?")) {
        return;
    }

    if (!assertWordsLeft("Say what?")) {
        return;
    }

    QString message = takeRest();

    player->talk(character, message);
}

File: data/commands/talkcommand.js
function TalkCommand() {
    Command.call(this);

    this.setDescription("Talk to a specific character who's in the same room. Everyone close to " +
                        "you can hear it.\n" +
                        "\n" +
                        "Example: talk earl Hey Earl, how are you?");
}

TalkCommand.prototype = new Command();
TalkCommand.prototype.constructor = TalkCommand;

TalkCommand.prototype.execute = function(player, command) {

    this.prepareExecute(player, command);

    var character = this.takeObject(player.currentRoom.characters);
    if (!this.requireSome(character, "Talk to who?")) {
        return;
    }

    if (!this.assertWordsLeft("Say what?")) {
        return;
    }

    var message = this.takeRest();

    player.talk(character, message);
};

CommandRegistry.registerCommand("talk", new TalkCommand());
